contracts/SQTGift.sol

```diff
@@ -0,0 +1,147 @@
+// Copyright (C) 2020-2023 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity 0.8.15;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721URIStorageUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721EnumerableUpgradeable.sol";
+import '@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol';
+
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import '@openzeppelin/contracts-upgradeable/utils/introspection/ERC165CheckerUpgradeable.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import './interfaces/ISQTGift.sol';
+
+contract SQTGift is Initializable, OwnableUpgradeable, ERC721Upgradeable, ERC721URIStorageUpgradeable, ERC721EnumerableUpgradeable, ISQTGift {
+  
+  address public sqtoken;
+  address public redeemer;
+
+  uint256 public maxSupply;
+  uint256 public maxSQT;
+
+  uint256 public defaultSQTValue;
+  uint256 public totalRedeemableSQT;
+  uint256 public totalRedeemedSQT;
+
+  RedeemRange[] public redeemRanges;
```

Comment:
I think the requirement may be not very clear. 
I'm thinking two type of Gift NFT. 1) use fixed amount 2) use random amount, decided when the token is claimed.
More things need to be taken care of when the amount is randomised. like the range of the result.
We may split them into two implementation for simplicity, or have a way to combine them into one contract.

Reply:
```
   function setRange(uint256 _minValue, uint256 _maxValue) public onlyOwner {
        minValue = _minValue;
        maxValue = _maxValue;
    }
        function randomValueInRange() private view returns (uint256) {
        // Simple random-like number generator. In production, consider a more secure way to generate randomness.
        uint256 random = uint256(keccak256(abi.encodePacked(block.timestamp, block.difficulty, msg.sender))) % (maxValue - minValue + 1);
        return minValue + random;
    }
    
```